internal/app/server/server.go

```diff
@@ -16,37 +16,101 @@ package server
 
 import (
 	"context"
+	"fmt"
 	"github.com/googleforgames/open-saves/internal/pkg/config"
+	"google.golang.org/grpc/health"
 	"net"
+	"os"
+	"os/signal"
+	"syscall"
 
 	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
 	"google.golang.org/grpc"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
 	"google.golang.org/grpc/reflection"
 
 	pb "github.com/googleforgames/open-saves/api"
 )
 
-// Run starts the Open Saves gRPC service.
+const serviceName = "Health"
+
+// Run starts the Open Saves and health check gRPC servers.
 func Run(ctx context.Context, network string, cfg *config.ServiceConfig) error {
-	log.Infof("starting server on %s %s", network, cfg.ServerConfig.Address)
+	// Handle servers lifecycle
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, os.Interrupt, syscall.SIGTERM)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g, ctx := errgroup.WithContext(ctx)
 
-	lis, err := net.Listen(network, cfg.ServerConfig.Address)
+	// gRPC Health Server
+	grpcHealthServer := grpc.NewServer()
+	healthServer := health.NewServer()
+	g.Go(func() error {
+		healthpb.RegisterHealthServer(grpcHealthServer, healthServer)
+
+		addr := cfg.HealthCheckConfig.Address
+		listener, err := net.Listen(network, addr)
+		if err != nil {
+			log.Errorf("gRPC Health server: failed to listen %s %s: %v", network, addr, err)
+			return err
+		}
+
+		log.Infof("gRPC health server serving at %s", addr)
+		return grpcHealthServer.Serve(listener)
+	})
+
+	// gRPC server
+	server, err := newOpenSavesServer(ctx, cfg)
 	if err != nil {
 		return err
 	}
-	defer func() {
-		if err := lis.Close(); err != nil {
-			log.Errorf("Failed to close %s %s: %v", network, cfg.ServerConfig.Address, err)
+	grpcServer := grpc.NewServer()
+
+	g.Go(func() error {
+		pb.RegisterOpenSavesServer(grpcServer, server)
+		reflection.Register(grpcServer)
+
+		addr := cfg.ServerConfig.Address
+		listener, err := net.Listen(network, addr)
+		if err != nil {
+			log.Errorf("gRPC server: failed to listen %s %s: %v", network, addr, err)
+			return err
 		}
-	}()
 
-	s := grpc.NewServer()
-	server, err := newOpenSavesServer(ctx, cfg)
+		log.Infof("starting server on %s %s", network, cfg.ServerConfig.Address)
+		healthServer.SetServingStatus(fmt.Sprintf("grpc.health.v1.%s", serviceName), healthpb.HealthCheckResponse_SERVING)
+		return grpcServer.Serve(listener)
+	})
+
+	select {
+	case <-sigs:
+		break
+	case <-ctx.Done():
+		break
+	}
+	log.Warn("received shutdown signal")
+
+	// Must be called as early as possible to stop accepting connections
+	cancel()
```

Comment:
Are we sure we want to cancel ongoing requests and transactions? I think "graceful shutdowns" usually means the server waits for any requests that are being processed within a reasonable timeouts. Apache and Nginx work like this IIRC.

Reply:
My understanding of cancel is that it does simply stops the service from accepting new connection. The graceful server shutdown takes care of waiting for the current requests to finish before quitting.

Reply:
I think that's what `GracefulStop()` does. Can you write a test for this behavior?

Reply:
The cancel is not needed / useful here according to this post with using errgroup (but it won't cause issues either): https://gist.github.com/akhenakh/38dbfea70dc36964e23acc19777f3869?permalink_comment_id=3139644#gistcomment-3139644 

For sure I will be testing this in our current set up with live traffic to make sure this is working as intended.

Reply:
cancel() is not necessary at all. I have added the initialization of the health check serving status.

Reply:
Your pull request seems based on the gist you put there. Do you have a license from the original author? I don't see the license in the gist that this was derived from.